File: CSLA10/CSLA/BusinessIdentity.vb
Imports System.Security.Principal
Imports System.Data.SqlClient

''' <summary>
''' 
''' </summary>
Namespace Security

  ''' <summary>
  ''' Implements a custom Identity class that supports
  ''' CSLA .NET data access via the DataPortal.
  ''' </summary>
  <Serializable()> _
  Public Class BusinessIdentity
    Inherits ReadOnlyBase

    Implements IIdentity

    Private mUsername As String
    Private mRoles As New ArrayList

#Region " IIdentity "

    ''' <summary>
    ''' Implements the IsAuthenticated property defined by IIdentity.
    ''' </summary>
    Public ReadOnly Property IsAuthenticated() As Boolean _
        Implements IIdentity.IsAuthenticated
      Get
        Return Len(mUsername) > 0
      End Get
    End Property

    ''' <summary>
    ''' Implements the AuthenticationType property defined by IIdentity.
    ''' </summary>
    Public ReadOnly Property AuthenticationType() As String _
        Implements IIdentity.AuthenticationType
      Get
        Return "CSLA"
      End Get
    End Property

    ''' <summary>
    ''' Implements the Name property defined by IIdentity.
    ''' </summary>
    Public ReadOnly Property Name() As String _
        Implements IIdentity.Name
      Get
        Return mUsername
      End Get
    End Property

#End Region

    Friend Function IsInRole(ByVal Role As String) As Boolean
      Return mRoles.Contains(Role)
    End Function

#Region " Create and Load "

    Friend Shared Function LoadIdentity(ByVal UserName As String, ByVal Password As String) As BusinessIdentity
      Return CType(DataPortal.Fetch(New Criteria(UserName, Password)), BusinessIdentity)
    End Function

    <Serializable()> _
    Private Class Criteria
      Public Username As String
      Public Password As String

      Public Sub New(ByVal Username As String, ByVal Password As String)
        Me.Username = Username
        Me.Password = Password
      End Sub
    End Class

    Private Sub New()
      ' prevent direct creation
    End Sub

#End Region

#Region " Data access "

    ''' <summary>
    ''' Retrieves the identity data for a specific user.
    ''' </summary>
    Protected Overrides Sub DataPortal_Fetch(ByVal Criteria As Object)
      Dim crit As Criteria = CType(Criteria, Criteria)

      mRoles.Clear()

      Dim cn As New SqlConnection(db("Security"))
      Dim cm As New SqlCommand()
      cn.Open()
      Try
        cm.Connection = cn
        cm.CommandText = "Login"
        cm.CommandType = CommandType.StoredProcedure
        cm.Parameters.Add("@user", crit.Username)
        cm.Parameters.Add("@pw", crit.Password)
        Dim dr As SqlDataReader = cm.ExecuteReader()
        Try
          If dr.Read() Then
            mUsername = crit.Username

            If dr.NextResult Then
              While dr.Read
                mRoles.Add(dr.GetString(0))
              End While
            End If

          Else
            mUsername = ""
          End If

        Finally
          dr.Close()
        End Try

      Finally
        cn.Close()
      End Try
    End Sub

#End Region

  End Class

End Namespace

File: cslacs10/CSLA/BusinessIdentity.cs
using System;
using System.Collections;
using System.Security.Principal;
using System.Data;
using System.Data.SqlClient;

namespace CSLA.Security
{
  /// <summary>
  /// Implements a custom Identity class that supports
  /// CSLA .NET data access via the DataPortal.
  /// </summary>
  [Serializable()]
  public class BusinessIdentity : ReadOnlyBase, IIdentity
  {
    string _username = string.Empty;
    ArrayList _roles = new ArrayList();

    #region IIdentity

    /// <summary>
    /// Implements the IsAuthenticated property defined by IIdentity.
    /// </summary>
    bool IIdentity.IsAuthenticated
    {
      get
      {
        return (_username.Length > 0);
      }
    }

    /// <summary>
    /// Implements the AuthenticationType property defined by IIdentity.
    /// </summary>
    string IIdentity.AuthenticationType
    {
      get
      {
        return "CSLA";
      }
    }

    /// <summary>
    /// Implements the Name property defined by IIdentity.
    /// </summary>
    string IIdentity.Name
    {
      get
      {
        return _username;
      }
    }

    #endregion

    internal bool IsInRole(string role)
    {
      return _roles.Contains(role);
    }

    #region Create and Load

    internal static BusinessIdentity LoadIdentity(
                              string userName, string password)
    {
      return (BusinessIdentity)DataPortal.Fetch(
        new Criteria(userName, password));
    }

    [Serializable()]
    private class Criteria
    {
      public string Username;
      public string Password;

      public Criteria(string username, string password)
      {
        Username = username;
        Password = password;
      }
    }

    private BusinessIdentity() {} // prevent direct creation

    #endregion

    #region Data access

    /// <summary>
    /// Retrieves the identity data for a specific user.
    /// </summary>
    protected override void DataPortal_Fetch(object criteria)
    {
      Criteria crit = (Criteria)criteria;

      _roles.Clear();

      SqlConnection cn = new SqlConnection(DB("Security"));
      SqlCommand cm = new SqlCommand();
      cn.Open();
      try
      {
        cm.Connection = cn;
        cm.CommandText = "Login";
        cm.CommandType = CommandType.StoredProcedure;
        cm.Parameters.Add("@user", crit.Username);
        cm.Parameters.Add("@pw", crit.Password);
        SqlDataReader dr = cm.ExecuteReader();
        try
        {
          if(dr.Read())
          {
            _username = crit.Username;

            if(dr.NextResult())
              while(dr.Read())
                _roles.Add(dr.GetString(0));
          }
          else
            _username = string.Empty;
        }
        finally
        {
          dr.Close();
        }
      }
      finally
      {
        cn.Close();
      }
    }

    #endregion

  }
}
